### ISSUE TITLE
RBAC

### ISSUE THREAD

**Issue by jesuslinares** [wazuh/wazuh-api#114]
Implement role-based access control (RBAC):
- Roles are created for several requests.
- The permissions to perform certain operations are assigned to specific roles.
- Users/groups are assigned to particular roles.

Concepts:
  - Secured Resource: A resource to which access is restricted. `Requests`.
  - Privilege: A named group of one or more actions that a user may execute against a secured resource. `GET, POST, PUT, DELETE`.
  - Permissions: A set of one or more privileges against a secured resource. `GET privelege on /agents request`.
  - Role: A named set of permission. `admin`.
  - User: The authenticated user. `user01`.
  - Group: One or more groups to which a user belongs. `soc-team`.

## Configuration example

`roles_mapping.yaml file`
```
admin:
  '*':
    - '*'
user:
  '/agents*':
    - 'GET'
    - 'POST'
    - 'PUT'
viewer:
  '/agents*':
    - 'GET'
```

The fields support wildcards (*), it will match until the first occurrence of the next character, if present.
Examples:
```
  '/*'
  '/agents/*'
  '/agents/1*'
  '/agents/name/type-*'
  '/agents/groups/web*'
  '/agents/groups/webservers/*'
  '/agents/groups/web*/'
```

It there are no wildcards, the match must be exact:
```
special-remove:
  '/agents?status=disconnected&older_than=15d'
    - DELETE
  '/agents?status=neverconnected&older_than=15d'
    - DELETE
```

## How to assign users and groups to a role
`roles_config.yaml file`
```
admin: 
  - "John" 
user:
  - "Jane" 
```

## Authentication

Secure the API with JSON Web Tokens.

## Next steps
- App integration (https://github.com/wazuh/wazuh-kibana-app/issues/619)
- Limit the responses to a group of agents
```
admin-webservers:
  '/agents':
    'methods':
      - 'GET'
      - 'POST'
      - 'PUT'
      - 'DELETE'
    'groups':
      - webservers
```
- Integration with LDAP/Active Directory

**Comment by juergenc** [comment]
Hello team and @jesusgn90 ,

referring to the [comment](https://github.com/wazuh/wazuh-api/issues/114#issuecomment-404211247) of @jesusgn90, persist also along with the token information the user-ID, user's source IP-asddress and the login attempts in order to recognise login brute force attacks and lock the user account or deny further login attempts for the next _n_ minutes.

**Comment by hsluoyz** [comment]
Hi guys, It looks like this issue has been two years and the RBAC feature is getting more and more complicated here. We are developing an authz library called Casbin and I think it may help you simplify the whole solution including the tables: https://casbin.org/

**Comment by davidjiglesias** [comment]
Already implemented in 4.0
